### PR TITLE
add subdivision option to extract_surface

### PR DESCRIPTION
**Pull request by akaszynski**
### Add subdivision

This PR adds in the missing ``SetNonlinearSubdivisionLevel`` option to ``extract_surface``.  It also adds in an additional example for the `extract_surface` filter.

```py
import pyvista as pv
import numpy as np

lin_pts = np.array([[-1, -1, -1],  # point 0
                    [ 1, -1, -1],  # point 1
                    [ 1,  1, -1],  # point 2
                    [-1,  1, -1],  # point 3
                    [-1, -1,  1],  # point 4
                    [ 1, -1,  1],  # point 5
                    [ 1,  1,  1],  # point 6
                    [-1,  1,  1]], np.double)  # point 7

# these are the "midside" points of a quad cell
quad_pts = np.array([
    (lin_pts[1] + lin_pts[0])/2.0,
    (lin_pts[1] + lin_pts[2])/2.0,
    (lin_pts[2] + lin_pts[3])/2.0,
    (lin_pts[3] + lin_pts[0])/2.0,
    (lin_pts[4] + lin_pts[5])/2.0,
    (lin_pts[5] + lin_pts[6])/2.0,
    (lin_pts[6] + lin_pts[7])/2.0,
    (lin_pts[7] + lin_pts[4])/2.0,
    (lin_pts[0] + lin_pts[4])/2.0,
    (lin_pts[1] + lin_pts[5])/2.0,
    (lin_pts[2] + lin_pts[6])/2.0,
    (lin_pts[3] + lin_pts[7])/2.0], np.double)

# introduce a minor variation to the location of the mid-side points
quad_pts += np.random.random(quad_pts.shape)*0.3
pts = np.vstack((lin_pts, quad_pts))

# create the grid

# If you are using vtk>=9, you do not need the offset array
offset = np.array([0])
cells = np.asarray(np.hstack((20, np.arange(20))), dtype=np.int64)
celltypes = np.array([VTK_QUADRATIC_HEXAHEDRON])
grid = pv.UnstructuredGrid(offset, cells, celltypes, pts)

surf_subdivided = grid.extract_surface(subdivision=5)
surf_subdivided.plot(show_scalar_bar=False)
```

image

### CODE REVIEW

**Comment by akaszynski** [comment]
Note to self to add subdivision test and example.

**Comment by akaszynski** [comment]
Ready for review.

**Comment by akaszynski** [comment]
@adeak, if you have a moment, could you take a look at this?

File: examples/01-filter/extract-surface.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+surf = grid.extract_surface()
+surf.plot(show_scalar_bar=False)
+
+
+###############################################################################
+# Quadratic Surface Subdivision
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Should your UnstructuredGrid contain quadratic cells, you can
+# generate a smooth surface based on the "shape functions" of each
+# cell.
```
Comment (adeak) [inline examples/01-filter/extract-surface.py:68]: This part and its relationship to the following code block was not clear to me at all. Let's see if I understand it correctly, having read both this and the corresponding docstring.

Quadratic cells have an inherent "shape function" which means that even though they are defined in terms of pointy cells (see e.g. [vtk figure](https://raw.githubusercontent.com/lorensen/VTKExamples/master/src/Testing/Baseline/Cxx/GeometricObjects/TestIsoparametricCellsDemo.png) and the first plot in this example) they have an underlying (biquadratic?) interpolator. The `subdivision` kwarg gives an appropriate linear approximation of this underlying smooth function. So for non-`UnstructuredGrid`s the whole `subdivision` thing is moot, because there's no interpolator (even though linear cells could be subdivided linearly, but presumably something like `.subdivide` is meant for that). And this sentence in the example is synonymous with the functionality of `extract_surface(subdivision=5)` even if at first it reads a bit like a _non sequitur_.

If this is the case, it might be worth making this clearer either here or in the docstring. Admittedly I don't know anything about quadratic cells, but the "shape function" mention here was missing a reference for me, and it might be good to mention the nonlinear underlying function in the docstring. I don't have a concrete suggestion on _how_ to improve this, though, I only have my initial confusion.

File: pyvista/core/filters.py [hunk 2]
```diff
@@ -2265,23 +2266,52 @@ def extract_surface(dataset, pass_pointid=True, pass_cellid=True, inplace=False)
             Adds a cell array "vtkOriginalPointIds" that idenfities which
             original cells these surface cells correspond to
 
+        subdivision : int, optional
+            If the input is an unstructured grid with nonlinear faces,
```
Comment (adeak) [inline pyvista/core/filters.py:2270]: Concerning my long-winded comment on the example: I guess this part is OK if we assume that the reader knows what "nonlinear faces" exactly entails. Arguably it's not the reference doc's job to teach basic VTK, so perhaps any further explanation is better left with the example.

File: pyvista/core/filters.py [hunk 3]
```diff
@@ -2265,23 +2266,52 @@ def extract_surface(dataset, pass_pointid=True, pass_cellid=True, inplace=False)
...
         if pass_pointid:
             surf_filter.PassThroughCellIdsOn()
         if pass_cellid:
             surf_filter.PassThroughPointIdsOn()
-        surf_filter.Update()
 
-        # TODO: Add
-        # surf_filter.SetNonlinearSubdivisionLevel(subdivision)
+        if subdivision != 1:
+            surf_filter.SetNonlinearSubdivisionLevel(subdivision)
```
Comment (adeak) [inline pyvista/core/filters.py:2309]: I guess the VTK name makes the functionality very clear. It would probably be more obvious if we also called the parameter `nonlinear_subdivision`, but that might be impractically unwieldy.

File: tests/plotting/test_plotting.py [hunk 4]
```diff
@@ -50,7 +50,7 @@
 
 # IMAGE warning/error thresholds (assumes using use_vtk)
 IMAGE_REGRESSION_ERROR = 500  # major differences
-IMAGE_REGRESSION_WARNING = 200  # minor differences
+IMAGE_REGRESSION_WARNING = 300  # minor differences
```
Comment (adeak) [inline tests/plotting/test_plotting.py:53]: I trust your judgement on this one ;)

File: tests/test_filters.py [hunk 5]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
     assert boundary.n_points
 
 
+def test_extract_surface():
+    # create a single hexahedral cell
```
Comment (adeak) [inline tests/test_filters.py:876]: Again redundancy for the sake of clarity might be better
```suggestion
    # create a single quadratic hexahedral cell
```

File: tests/test_filters.py [hunk 6]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
...
+        (lin_pts[2] + lin_pts[3])/2.0,
+        (lin_pts[3] + lin_pts[0])/2.0,
+        (lin_pts[4] + lin_pts[5])/2.0,
+        (lin_pts[5] + lin_pts[6])/2.0,
+        (lin_pts[6] + lin_pts[7])/2.0,
+        (lin_pts[7] + lin_pts[4])/2.0,
+        (lin_pts[0] + lin_pts[4])/2.0,
+        (lin_pts[1] + lin_pts[5])/2.0,
+        (lin_pts[2] + lin_pts[6])/2.0,
+        (lin_pts[3] + lin_pts[7])/2.0], np.double)
```
Comment (adeak) [inline tests/test_filters.py:898]: See remark for the example.

File: tests/test_filters.py [hunk 7]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
...
+
+    # introduce a minor variation to the location of the mid-side points
+    quad_pts += np.random.random(quad_pts.shape)*0.25
```
Comment (adeak) [inline tests/test_filters.py:901]: It's probably not recommended to have random numbers in a test, but I can't think of an edge case where this could be problematic (relative shifts are pretty small).

It would also be nice to use the new random API (added in numpy 1.17 I think) for modernness' sake, but again that's not possible due to legacy versions.

File: tests/test_filters.py [hunk 8]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
...
+    pts = np.vstack((lin_pts, quad_pts))
+
+    cells = np.asarray([[20, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]], dtype=np.int64)
+    celltypes = np.array([VTK_QUADRATIC_HEXAHEDRON])
+    if pyvista._vtk.VTK9:
+        grid = pyvista.UnstructuredGrid(cells, celltypes, pts)
+    else:
+        grid = pyvista.UnstructuredGrid(np.array([0]), cells, celltypes, pts)
+
+    # expect each face to be divided 6 times since's it has a midside node
```
Comment (adeak) [inline tests/test_filters.py:911]:
```suggestion
    # expect each face to be divided 6 times since it has a midside node
```

File: tests/test_filters.py [hunk 9]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
...
+    surf = grid.extract_surface()
+    assert surf.n_faces == 36
+
+    # expect each face to be divided several more times than the linear extraction
+    surf_subdivided = grid.extract_surface(subdivision=5)
+    assert surf_subdivided.n_faces > surf.n_faces
+
```
Comment (adeak) [inline tests/test_filters.py:918]: Do we also want to test `subdivision=0`? In terms of code that would only test VTK, but we do specify this case in our docstring.

File: tests/test_filters.py [hunk 10]
```diff
@@ -871,6 +872,51 @@ def test_decimate_boundary():
...
+        (lin_pts[0] + lin_pts[4])/2.0,
+        (lin_pts[1] + lin_pts[5])/2.0,
+        (lin_pts[2] + lin_pts[6])/2.0,
+        (lin_pts[3] + lin_pts[7])/2.0], np.double)
+
+    # introduce a minor variation to the location of the mid-side points
+    quad_pts += np.random.random(quad_pts.shape)*0.25
+    pts = np.vstack((lin_pts, quad_pts))
+
+    cells = np.asarray([[20, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]], dtype=np.int64)
```
Comment (adeak) [inline tests/test_filters.py:904]: 116 characters, might be better to split it somewhere. First arg ends at 101 so it might be enough to split before `dtype`.

**Comment by adeak** [comment]
@akaszynski not sure if this is just lower priority or you missed the review, so ping.

File: examples/01-filter/extract-surface.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+# introduce a minor variation to the location of the mid-side points
+quad_pts += np.random.random(quad_pts.shape)*0.3
+pts = np.vstack((lin_pts, quad_pts))
+
+# create the grid
+
+# If you are using vtk>=9, you do not need the offset array
+offset = np.array([0])
+cells = np.asarray(np.hstack((20, np.arange(20))), dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:54]: If this was a huge array, the top would be faster as there's only one array generation, whereas an `astype` will always copy:
```py
In [1]: import numpy as np
In [2]: arr = np.ones(10000000, np.int32)
In [3]: timeit arr.astype(np.int32)
7.19 ms ± 16.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Your suggestion reads easier, so I'll just implement it.

[the remark this answers is not in the document]
Comment (adeak) [reply examples/01-filter/extract-surface.py:54]: Huh, I would've sworn that `.astype` also returns the array (or at most a view)! Thanks. I'll try to figure out if there's a way to have our cake and eat it too, this bothers me now.

[the remark this answers is not in the document]
Comment (adeak) [reply examples/01-filter/extract-surface.py:54]: OK, this was easy: `.astype(np.int64, copy=False)` :D

File: examples/01-filter/extract-surface.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+celltypes = np.array([VTK_QUADRATIC_HEXAHEDRON])
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:55]: Sadly, yes.  I need to do the same thing I just did for #1213 to the `UnstructuredGrid` class.  We check if all args are `np.ndarray` in:
https://github.com/pyvista/pyvista/blob/07d2fdd0ae655ae3aecf477c84098929effbb4f2/pyvista/core/pointset.py#L593-L594

This will have to be fixed in a separate PR.

File: examples/01-filter/extract-surface.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+import numpy as np
+import pyvista as pv
+from vtk import VTK_QUADRATIC_HEXAHEDRON
+
+###############################################################################
+# Create a quadratic cell and extract its surface
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Here we create a single hexahedral cell and then extract its surface
+# to demonstrate how to extract the surface of a UnstructuredGrid.
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:19]: Agreed.

File: examples/01-filter/extract-surface.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:54]: These are things I usually just gloss over when I'm being a lazy coder:
```py
In [5]: timeit arr.astype(np.int32, copy=False)
445 ns ± 9.16 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [6]: timeit arr.astype(np.int64, copy=False)
10.6 ms ± 165 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Seems that it only copies when necessary.

File: examples/01-filter/extract-surface.py (hunk 1 above)
Comment (akaszynski) [inline examples/01-filter/extract-surface.py:68]:
```suggestion
# generate a smooth surface based on the position of the 
"mid-edge" nodes.  This allows the plotting of cells
containing curvature.  For additional reference, please see:
https://prod.sandia.gov/techlib-noauth/access-control.cgi/2004/041617.pdf
```

File: examples/01-filter/extract-surface.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (adeak) [reply examples/01-filter/extract-surface.py:54]: Well I had no idea that keyword was a thing. We both learned something :)

File: examples/01-filter/extract-surface.py (hunk 1 above)
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:68]: I'm missing context here for sure.  It's too much to explain in a basic example, but I can give a two sentence explanation and a reference for those who want to read more.  For your edification, "quadratic elements" are widely used in finite element analysis to compute better approximations of static/harmonic characteristics of digital models.  There's two ways of doing this: denser models or models with more integration points per element.  Here's a good reference:
https://deust.wordpress.com/2014/11/30/h-method-p-method/#:~:text=This%20method%20refers%20to%20increasing,method%20uses%20few%20complex%20elements.

The cool part for us is we can visualize the geometric curvature inherent in "higher order" elements when the points along the edges are not placed directly between their corresponding edge nodes.

File: pyvista/core/filters.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/core/filters.py:2270]: I'm afraid that it's going to be an "exercise left up to the reader".  VTK docs contain some decent explanation, Paraview has good examples with pictures, and for the rest there's google.  I've included a better explanation in the examples docs.

File: examples/01-filter/extract-surface.py (hunk 1 above)
Comment (adeak) [reply examples/01-filter/extract-surface.py:68]: Looks like a good reference, thanks! Hopefully the link won't be too mobile in the future.

File: tests/plotting/test_plotting.py (hunk 4 above)
Comment (akaszynski) [reply tests/plotting/test_plotting.py:53]: This was for a separate issue that I shouldn't have done here.  I'll save this for another PR.

Comment (akaszynski) [inline tests/plotting/test_plotting.py:53]:
```suggestion
IMAGE_REGRESSION_WARNING = 200  # minor differences
```

File: examples/01-filter/extract-surface.py (hunk 1 above)
Comment (adeak) [reply examples/01-filter/extract-surface.py:68]: Also thanks for the explanation (it didn't load before I left my previous comment). I had very limited exposure to finite elements and I haven't heard about nonlinear elements. They indeed look useful :)

File: tests/test_filters.py (hunk 7 above)
Comment (akaszynski) [reply tests/test_filters.py:901]: We use `hypothesis` quite a bit, and haven't had issues with random values.  I was honestly just being lazy here, but I think it works better than including an array.

File: examples/01-filter/extract-surface.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    (lin_pts[2] + lin_pts[3])/2.0,
+    (lin_pts[3] + lin_pts[0])/2.0,
+    (lin_pts[4] + lin_pts[5])/2.0,
+    (lin_pts[5] + lin_pts[6])/2.0,
+    (lin_pts[6] + lin_pts[7])/2.0,
+    (lin_pts[7] + lin_pts[4])/2.0,
+    (lin_pts[0] + lin_pts[4])/2.0,
+    (lin_pts[1] + lin_pts[5])/2.0,
+    (lin_pts[2] + lin_pts[6])/2.0,
+    (lin_pts[3] + lin_pts[7])/2.0], np.double)
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/extract-surface.py:44]: A bunch of this is redundant.  Fixing...

File: pyvista/core/filters.py (hunk 3 above)
Comment (akaszynski) [reply pyvista/core/filters.py:2309]: Zen of Python tells me that this is a balance between:
```
Explicit is better than implicit.
Simple is better than complex.
```
However, considering that those not well versed in quadratic elements might not understand how you can "subdivide" the surface of an element, or even why you' want to do it here, I think you're right.  I'll implement it.

Comment (adeak) [reply pyvista/core/filters.py:2309]: Yeah, I'm on the fence myself.

File: tests/test_filters.py (hunk 10 above)
Comment (akaszynski) [reply tests/test_filters.py:904]: Just copied the example as it fits on one line.  I breakout in a cold sweat every time I go over 79 characters and am physically ill past 100 as my lines start wrapping in emacs.

File: tests/test_filters.py (hunk 9 above)
Comment (akaszynski) [reply tests/test_filters.py:918]: I literally have 300 lines of code in a closed source project to trick my module to return this result.  I've added the test here and I'll refactor my internal code.  Learn something every day...

File: examples/01-filter/extract-surface.py [hunk 15]
```diff
@@ -61,13 +63,13 @@
 
 
 ###############################################################################
-# Quadratic Surface Subdivision
+# Nonlinear Surface Subdivision
```
Comment (adeak) [inline examples/01-filter/extract-surface.py:66]: Great idea!

File: pyvista/core/filters.py [hunk 16]
```diff
@@ -2265,10 +2266,36 @@ def extract_surface(dataset, pass_pointid=True, pass_cellid=True, inplace=False)
             Adds a cell array "vtkOriginalPointIds" that idenfities which
             original cells these surface cells correspond to
```
Comment (adeak) [inline pyvista/core/filters.py:2267]: Unrelated to the PR but two typos, second one also present in previous parameter's description due to copy-paste. Also missing ending period, again for the previous parameter too.
```suggestion
            Adds a cell array "vtkOriginalCellIds" that identifies which
            original cells these surface cells correspond to.
```

**Review by adeak — APPROVED** [review]
Culombo-style "one more thing", but LGTM, thanks.
